unittests/forked_tests_savanna.cpp

```diff
@@ -0,0 +1,731 @@
+   //
+   // _nodes[3] is still having just produced the 2nd block by "sam", and with the `producers` schedule.
+   // skip 22 blocks in the future so that "pam" produces
+   auto node3_head = _nodes[3].produce_block(_block_interval_us * 22);
+   BOOST_REQUIRE_EQUAL(node3_head->producer, producers[1]);    // should be sam's last block
+   push_block(0, node3_head);
+   BOOST_REQUIRE_EQUAL(_nodes[3].head().id(), _nodes[0].head().id());  // fork switch on 1st block because of later timestamp
+   BOOST_REQUIRE_EQUAL(_nodes[3].head().id(), _nodes[1].head().id());  // push_block() propagated on peer which also fork switched
+
+   sb = _nodes[3].produce_block();
+   BOOST_REQUIRE_EQUAL(sb->producer, producers[2]);            // just switched to "pam"
+   sb = _nodes[3].produce_blocks(12);                          // after 12 blocks, should have switched to "dan"
+   BOOST_REQUIRE_EQUAL(sb->producer, producers[0]);            // chack that this is the case
+
+   push_blocks(3, 0, node3_head->block_num() + 1);             // push the last 13 produced blocks to _nodes[0]
+   BOOST_REQUIRE_EQUAL(_nodes[0].head().id(), _nodes[3].head().id());  // _nodes[0] caught up
+   BOOST_REQUIRE_EQUAL(_nodes[1].head().id(), _nodes[3].head().id());  // _nodes[0] peer was updated as well
+
+   // unsplit the network
+   set_partition({});
+
+   // produce an even more recent block on _nodes[0] so that it will be the uncontested head
+   sb = _nodes[0].produce_block(_block_interval_us, true);     // no_throw = true because of expired transaction
+   BOOST_REQUIRE_EQUAL(_nodes[0].head().id(), _nodes[2].head().id());
+   BOOST_REQUIRE_EQUAL(_nodes[0].head().id(), _nodes[3].head().id());
+
+   // and verify lib advances.
+   auto lib = _nodes[0].lib_block->block_num();
+   size_t tries = 0;
+   while (_nodes[0].lib_block->block_num() <= lib + 3 && ++tries < 10) {
+      _nodes[0].produce_block();
+   }
+   BOOST_REQUIRE_GT(_nodes[0].lib_block->block_num(), lib + 3);
+   BOOST_REQUIRE_EQUAL(_nodes[0].lib_block->block_num(), _nodes[3].lib_block->block_num());
+} FC_LOG_AND_RETHROW()
+
+// ---------------------------- verify_savanna_fork_choice --------------------------
+// Verify fork choice criteria for Savanna:
+//   last_final_block_num > last_qc_block_num > timestamp
+//
+// - Simulate two network partitions: P0 (node {0}) and P1 (nodes {1, 2, 3}).
+// - produce 2 blocks on P1. finality will advance by 2 blocks.
+// - produce 2 blocks on P0. finality will advance by 1 block only (because no quorum)
+//   but they'll have a later timestamp
+// - push the 2 newly produced blocks from P1 to P0
+// - check that we fork-switched to P1's head (despite P0's head timestamp being later)
+// - Unpartition the network, veerify lib advances.
+// ----------------------------------------------------------------------------------
+BOOST_FIXTURE_TEST_CASE( verify_savanna_fork_choice, savanna_cluster::cluster_t) try {
+   while (_nodes[0].head().block_num() < 3) {
```

Comment:
How about this while loop?

Reply:
Same, not necessary, I removed it. I like making tests as simple as possible, thanks for the help!